src/main/java/fi/aalto/cs/apluscourses/utils/ConsList.java

```diff
@@ -0,0 +1,33 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import java.util.AbstractList;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Lisp-like list implementation.
+ * Inspired by https://stackoverflow.com/a/2634421
+ *
+ * @param <T> Type of list elements.
+ */
+public class ConsList<T> extends AbstractList<T> {
```

Comment:
could you please elaborate on what is a benefit of this approach over the simple creation of a new `List` with that desired element as (to be added) as the first one?

Reply:
You mean like creating a new `ArrayList` or `LinkedList` or other standard implementations? The problem with those is that a creating a new list based on an old one requires all the elements to be copied from the old to the new (in case of `ArrayList`, even extra copying is needed to insert an element in the beginning of the list).

This class is used in recursive computation where each step produces a new list with a new item in the beginning of the list. If we used, say Java's `LinkedList`, we would be copying items again and again from a temporary list to another which would lead to quite a rampant and difficult-to-predict time complexity (possibly memory complexity too) .

It's quite surprising that Java's standard library doesn't have a class like this ready-made. On the other hand, the code that uses this, is not very Java-like but following the patterns of functional programming languages such as Lisp. It could be refactored to carry out the computation in more Java-like (imperative) manner but it would then be more difficult to understand it and reason about its correctness.

Reply:
alright, your answer helped me to (find more details) understand! :bow:

Reply:
A cons list has horrible performance when iterating (compared to `ArrayList`), but the recursive solution in `traverseAndFind` with cons list is very neat indeed

Reply:
Right. It should have similar performance in iteration and random-access as `LinkedList` (well, it is a linked list). Which is of course poor compared to arrays.